File: backend-comic/src/main/java/com/thjvjpxx/backend_comic/service/VipPackageService.java
package com.thjvjpxx.backend_comic.service;

import java.util.List;

import com.thjvjpxx.backend_comic.dto.request.VipPackageRequest;
import com.thjvjpxx.backend_comic.dto.response.BaseResponse;
import com.thjvjpxx.backend_comic.model.VipPackage;

public interface VipPackageService {

    /**
     * Lấy danh sách tất cả gói VIP (có phân trang và tìm kiếm)
     * 
     * @param page     Trang hiện tại
     * @param limit    Số lượng mỗi trang
     * @param search   Từ khóa tìm kiếm
     * @param isActive Trạng thái hoạt động
     * @return Response chứa danh sách gói VIP
     */
    BaseResponse<List<VipPackage>> getAllVipPackages(int page, int limit, String search, Boolean isActive);

    /**
     * Lấy gói VIP theo ID
     * 
     * @param id ID gói VIP
     * @return Response chứa gói VIP
     */
    BaseResponse<VipPackage> getVipPackageById(String id);

    /**
     * Tạo gói VIP mới
     * 
     * @param request Dữ liệu gói VIP
     * @return Response chứa gói VIP đã tạo
     */
    BaseResponse<VipPackage> createVipPackage(VipPackageRequest request);

    /**
     * Cập nhật gói VIP
     * 
     * @param id      ID gói VIP
     * @param request Dữ liệu gói VIP
     * @return Response chứa gói VIP đã cập nhật
     */
    BaseResponse<VipPackage> updateVipPackage(String id, VipPackageRequest request);

    /**
     * Xóa gói VIP (soft delete - chuyển thành inactive)
     * 
     * @param id ID gói VIP
     * @return Response chứa gói VIP đã xóa
     */
    BaseResponse<VipPackage> deleteVipPackage(String id);

    /**
     * Xóa hoàn toàn gói VIP khỏi database
     * 
     * @param id ID gói VIP
     * @return Response chứa gói VIP đã xóa
     */
    BaseResponse<String> permanentDeleteVipPackage(String id);

    /**
     * Lấy danh sách gói VIP để hiển thị cho khách hàng (chỉ active)
     * 
     * @return Response chứa danh sách gói VIP
     */
    BaseResponse<List<VipPackage>> getPublicVipPackages();
}

File: frontend-comic/src/services/vipPackageService.ts
import { VipPackage, VipPackageCreateUpdate } from "@/types/vipPackage";
import { fetchApi } from "./api";

export const getAllVipPackages = async (
  page: number = 1,
  limit: number = 5,
  search?: string,
  status?: boolean
) => {
  let endpoint = `/vip-packages?page=${page}&limit=${limit}`;

  if (search) {
    endpoint += `&search=${encodeURIComponent(search)}`;
  }

  if (status !== undefined) {
    endpoint += `&isActive=${status}`;
  }

  return await fetchApi<VipPackage[]>(endpoint);
}

export const createVipPackage = async (data: VipPackageCreateUpdate) => {
  const endpoint = `/vip-packages`;

  return await fetchApi<VipPackage>(endpoint, {
    method: "POST",
    data: data,
  });
}

export const updateVipPackage = async (id: string, data: VipPackageCreateUpdate) => {
  const endpoint = `/vip-packages/${id}`;

  return await fetchApi<VipPackage>(endpoint, {
    method: "PUT",
    data: data,
  });
}

export const deleteVipPackage = async (id: string) => {
  const endpoint = `/vip-packages/${id}`;

  return await fetchApi<VipPackage>(endpoint, {
    method: "DELETE",
  });
}

export const permanentDeleteVipPackage = async (id: string) => {
  const endpoint = `/vip-packages/${id}/permanent`;

  return await fetchApi<VipPackage>(endpoint, {
    method: "DELETE",
  });
}

export const getPublicVipPackages = async () => {
  const endpoint = `/public/vip-packages`;

  return await fetchApi<VipPackage[]>(endpoint);
}